File: sdk/nodejs/src/tier2.ts
import * as crypto from 'crypto';
import { Tier1 } from './tier1';
import { C2CPOFormatViolationError } from './exceptions';

/**
 * Tier 2 — Custom Value Encoding.
 *
 * Extends Tier 1 by also obfuscating field *values*:
 * - **Integers** → Base36 uppercase string (e.g. 1000 → "RS")
 * - **Strings**  → 12-character uppercase HMAC-SHA256 token
 *
 * Requires a shared `masterSecret` (32+ bytes) on both producer and consumer.
 * Load from an environment variable or KMS backend — never hard-code.
 *
 * Skilled-attacker barrier: ~3–8 hours (vs. ~1–3 hours for Tier 1).
 * Latency: < 2ms for typical payloads.
 *
 * @example
 * ```typescript
 * import { Tier2, KeyManager } from 'c2cpo';
 *
 * const secret = await new KeyManager().getMasterKey();
 * const encoder = new Tier2('payment-service', secret);
 * const encoded = encoder.encode({ amount: 1000, user_id: 'usr_123' });
 * // → { XR7F2K9Q: 'RS', U9B3MX1P: '4A2B8C1D9E3F' }
 * ```
 */
export class Tier2 extends Tier1 {
  protected masterSecret: Buffer;

  protected get tierNumber(): number { return 2; }

  /**
   * @param endpointId   Unique service endpoint identifier.
   * @param masterSecret Shared 32-byte secret key (load via KeyManager).
   * @param webhookUrl   Optional SIEM webhook URL for violation events.
   */
  constructor(endpointId: string, masterSecret: Buffer, webhookUrl?: string) {
    super(endpointId, webhookUrl);
    this.masterSecret = masterSecret;
    this.anomalyDetector.tier = 2;
  }

  /** Encode integer as Base36 uppercase string. */
  protected encodeBase36(num: number): string {
    if (num === 0) return '0';
    const isNegative = num < 0;
    let n = Math.abs(num);
    const alphabet = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ';
    let base36 = '';
    while (n > 0) {
      base36 = alphabet[Math.floor(n % 36)] + base36;
      n = Math.floor(n / 36);
    }
    return isNegative ? '-' + base36 : base36;
  }

  /** Decode Base36 string back to integer. */
  protected decodeBase36(base36: string): number {
    return parseInt(base36, 36);
  }

  /** Replace a string with a 12-char uppercase HMAC-SHA256 token. */
  protected encodeString(value: string): string {
    const h = crypto.createHmac('sha256', this.masterSecret);
    h.update(value, 'utf8');
    return h.digest('hex').substring(0, 12).toUpperCase();
  }

  /** Encode payload with Tier 1 field obfuscation + Tier 2 value encoding. */
  public encode(payload: Record<string, any>): Record<string, any> {
    const encodedPayload: Record<string, any> = {};

    for (const [key, value] of Object.entries(payload)) {
      const encodedKey = Tier1.EPOCH_0_MAPPING[key] || key;

      let encodedValue = value;
      if (typeof value === 'number' && Number.isInteger(value)) {
        encodedValue = this.encodeBase36(value);
      } else if (typeof value === 'string') {
        encodedValue = this.encodeString(value);
      }

      encodedPayload[encodedKey] = encodedValue;
    }

    return encodedPayload;
  }

  /**
   * Decode a Tier 2 encoded payload.
   * Integer values (Base36) are decoded back to numbers.
   * String values (HMAC tokens) are returned as-is.
   */
  public decode(
    payload: Record<string, any>,
    rawRequestStr: string,
    sourceIp: string = 'unknown'
  ): Record<string, any> {
    const decodedPayload: Record<string, any> = {};
    let violationDetected = false;

    for (const [key, value] of Object.entries(payload)) {
      if (key in Tier1.EPOCH_0_MAPPING) {
        violationDetected = true;
      }

      const decodedKey = this.reversedMapping[key];
      if (decodedKey) {
        let decodedValue = value;
        if (typeof value === 'string' && decodedKey === 'amount') {
          const parsed = this.decodeBase36(value);
          if (!isNaN(parsed)) {
            decodedValue = parsed;
          }
        }
        decodedPayload[decodedKey] = decodedValue;
      } else if (!(key in Tier1.EPOCH_0_MAPPING)) {
        decodedPayload[key] = value;
      }
    }

    if (violationDetected) {
      this.anomalyDetector.logFormatViolation(rawRequestStr, 0, sourceIp);
      throw new C2CPOFormatViolationError(
        'FORMAT_VIOLATION: Unencoded standard fields detected in payload.'
      );
    }

    return decodedPayload;
  }
}

File: sdk/python/c2cpo/tier2.py
from typing import Dict, Any, Union
import hmac
import hashlib
from .tier1 import Tier1
from .exceptions import C2CPOFormatViolationError


class Tier2(Tier1):
    """
    Tier 2 — Custom Value Encoding.

    Extends Tier 1 by also obfuscating field *values*:

    - **Integers** are encoded as Base36 uppercase strings (e.g. 1000 → "RS")
    - **Strings** are replaced by a 12-character uppercase HMAC-SHA256 token
      derived from the master secret and the original string value

    Requires a shared ``master_secret`` (32+ bytes) on both producer and
    consumer.  The master secret should be loaded from an environment variable
    or a KMS backend via :class:`KeyManager` — never hard-coded.

    Skilled-attacker barrier: ~3–8 hours (vs. ~1–3 hours for Tier 1).
    Latency overhead: < 2ms for typical payloads.

    Usage::

        import os
        from c2cpo import Tier2, KeyManager

        key_manager = KeyManager()
        secret = key_manager.get_master_key()

        encoder = Tier2(endpoint_id="payment-service", master_secret=secret)
        encoded = encoder.encode({"amount": 1000, "user_id": "usr_123"})
        # → {"XR7F2K9Q": "RS", "U9B3MX1P": "4A2B8C1D9E3F"}

        decoder = Tier2(endpoint_id="payment-service", master_secret=secret)
        decoded = decoder.decode(encoded, raw_request_str="{...}")
        # → {"amount": 1000, "user_id": "4A2B8C1D9E3F"}
        # Note: string values decode statelessly — original value not recovered
    """

    _TIER_NUMBER: int = 2

    def __init__(self, endpoint_id: str, master_secret: bytes, webhook_url: str = None):
        """
        Args:
            endpoint_id:   Unique identifier for this service endpoint.
            master_secret: Shared 32-byte (256-bit) secret key.
                           Load via KeyManager — never hard-code.
            webhook_url:   Optional SIEM webhook URL for violation events.
        """
        super().__init__(endpoint_id=endpoint_id, webhook_url=webhook_url)
        self.master_secret = master_secret
        self.anomaly_detector.tier = 2

    # --- Value Encoders ---

    def _encode_base36(self, number: int) -> str:
        """Encode an integer as a Base36 uppercase string."""
        if number == 0:
            return '0'
        alphabet = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ'
        base36 = ''
        sign = '-' if number < 0 else ''
        number = abs(number)
        while number:
            number, i = divmod(number, 36)
            base36 = alphabet[i] + base36
        return sign + base36

    def _decode_base36(self, base36: str) -> int:
        """Decode a Base36 string back to an integer."""
        return int(base36, 36)

    def _encode_string(self, value: str) -> str:
        """
        Replace a string value with a 12-character HMAC-SHA256 token.

        HMAC is keyed with the master secret so the token is deterministic
        for the same input but unguessable without K.
        """
        h = hmac.new(self.master_secret, value.encode('utf-8'), hashlib.sha256).hexdigest()
        return h[:12].upper()

    def encode(self, payload: Dict[str, Any]) -> Dict[str, Any]:
        """
        Encode payload with Tier 1 field-name obfuscation and Tier 2
        value encoding (Base36 for integers, HMAC token for strings).

        Args:
            payload: Dict with standard field names and plain values.

        Returns:
            Dict with obfuscated keys and encoded values.
        """
        encoded_payload = {}
        for key, value in payload.items():
            encoded_key = self.EPOCH_0_MAPPING.get(key, key)

            if isinstance(value, int):
                encoded_value = self._encode_base36(value)
            elif isinstance(value, str):
                encoded_value = self._encode_string(value)
            else:
                encoded_value = value

            encoded_payload[encoded_key] = encoded_value
        return encoded_payload

    def decode(
        self,
        payload: Dict[str, Any],
        raw_request_str: str,
        source_ip: str = "unknown",
    ) -> Dict[str, Any]:
        """
        Decode a Tier 2 encoded payload.

        Integer values (stored as Base36 strings) are decoded back to int.
        String values (stored as HMAC tokens) are returned as-is — the
        original string is not recoverable without the plaintext (by design).

        Raises:
            C2CPOFormatViolationError: If unencoded standard fields detected.
        """
        decoded_payload = {}
        violation_detected = False

        for key, value in payload.items():
            if key in self.EPOCH_0_MAPPING:
                violation_detected = True

            decoded_key = self.reversed_mapping.get(key)
            if decoded_key:
                if isinstance(value, str):
                    try:
                        if decoded_key == "amount":
                            decoded_value = self._decode_base36(value)
                        else:
                            decoded_value = value
                    except ValueError:
                        decoded_value = value
                else:
                    decoded_value = value
                decoded_payload[decoded_key] = decoded_value
            elif key not in self.EPOCH_0_MAPPING:
                decoded_payload[key] = value

        if violation_detected:
            self.anomaly_detector.log_format_violation(
                raw_request=raw_request_str,
                epoch_number=0,
                source_ip=source_ip,
            )
            raise C2CPOFormatViolationError(
                "FORMAT_VIOLATION: Unencoded standard fields detected in payload."
            )

        return decoded_payload
